Add new post processor googlecompute-import

seanmalloy opened the pull request:
Add new post processor googlecompute-import. Allows for automating the GCE image import process using Packer.

https://cloud.google.com/compute/docs/images/import-existing-image

seanmalloy commented:
This code only works successfully when compiled with Go 1.9. It does work when compiled with Go 1.10. I have not had a chance to determine the exact cause of the problem yet. I'm suspecting that this is a bug in Go 1.10.

Details of the issue that I know so far ...

When compiled with Go 1.10 GCP does not recognize the tar.gz file created by the compress post-processor as a valid TAR file. The GCE image creation API call fails.

In the near future I will be digging into this issue to determine the exact cause of this problem. In the meantime any feedback on this PR would be greatly appreciated.

mwhooker commented:
@seanmalloy Go changed the way tar works in 1.10. It may be that it uses a different format now that gcp doesn't recognize. See https://golang.org/pkg/archive/tar/#Format. You might be able to fix it by setting the format explicitly, but you'll need to add a special code path for go 1.9 builds.

see https://github.com/hashicorp/packer/pull/6084 for another issue dealing with tar on go 1.10

rickard-von-essen commented on post-processor/googlecompute-import/post-processor.go:
```diff
@@ -0,0 +1,204 @@
+	ImageName         string            `mapstructure:"image_name"`
+	ProjectId         string            `mapstructure:"project_id"`
+	AccountFile       string            `mapstructure:"account_file"`
+	KeepOriginalImage bool              `mapstructure:"keep_input_artifact"`
```
This should not be here, see https://github.com/hashicorp/packer/blob/master/post-processor/vagrant/post-processor.go#L153 how this should work.

seanmalloy replied:
I believe you are asking me to get rid of the `keep_input_artifact` configuration option and instead use  `provider.KeepInputArtifact()`. I'm not really super good and Go Lang, so ...

Do I need to implement the entire `PostProcessProvider` function?

https://github.com/hashicorp/packer/blob/master/post-processor/vagrant/post-processor.go#L75

rickard-von-essen replied:
Sorry, I'm wrong, you can just ignore this comment.

rickard-von-essen commented on post-processor/googlecompute-import/post-processor.go:
```diff
@@ -0,0 +1,204 @@
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	Bucket            string            `mapstructure:"bucket"`
```
Maybe there should be something like: https://www.packer.io/docs/post-processors/amazon-import.html#s3_key_name

seanmalloy replied:
Partially fixed. I added a new config option `gcs_object_name`, but the time stamp in the default value is not being interpolated. I'm getting a file literally named `packer-import-{{timestamp}}.tar.gz` in the GCS bucket. I'll need to sort this out.

rickard-von-essen commented on post-processor/googlecompute-import/post-processor.go:
```diff
@@ -0,0 +1,204 @@
+		err := fmt.Errorf(
+			"incompatible artifact type: %s\nCan only import from Compress post-processor artifacts",
+			artifact.BuilderId())
+		return nil, p.config.KeepOriginalImage, err
```
For simplicity "keep artefact" can be `false` when failing.

seanmalloy replied:
Fixed

seanmalloy commented:
@mwhooker the GCE import documentation explicitly states that the GNU tar format must be used. Looks like i will need to apply the fixes from #6084 and #6085  to the compress post-processor.

I'm glad this is a Go 1.10 feature and not a bug. Thanks for pointing me in the right direction on this.

seanmalloy commented on website/source/docs/post-processors/googlecompute-import.html.md:
```diff
@@ -42,6 +42,8 @@ for details.
 
 ### Optional
 
+-   `gcs_object_name` (string) - Tne name of the GCS object in `bucket` where the RAW disk image will be copied for import. Defaults to "packer-import-{{timestamp}}.tar.gz".
```
Fixed

seanmalloy commented on post-processor/googlecompute-import/post-processor.go:
```diff
@@ -47,13 +47,29 @@ func (p *PostProcessor) Configure(raws ...interface{}) error {
 		InterpolateContext: &p.config.ctx,
+		InterpolateFilter: &interpolate.RenderFilter{
+			Exclude: []string{
+				"gcs_object_name",
```
I could not get this working with doing a explicit interpolation. It looks like this is the same way the amazon-import post-processor works too.

https://github.com/hashicorp/packer/blob/master/post-processor/amazon-import/post-processor.go#L52
https://github.com/hashicorp/packer/blob/master/post-processor/amazon-import/post-processor.go#L64
https://github.com/hashicorp/packer/blob/master/post-processor/amazon-import/post-processor.go#L109

Maybe there is a better way to do this, but it seems like the only way I could get it to work correctly was to do it the same way the amazon-import post processor does it

seanmalloy commented:
@rickard-von-essen I believe I have addressed all of your review comments. Let me know if any other changes are needed.

My next step is to get the compress post-processor using the GNU tar format with Go 1.10.

rickard-von-essen commented on post-processor/googlecompute-import/post-processor.go:
```diff
@@ -0,0 +1,225 @@
+	ImageName         string            `mapstructure:"image_name"`
+	ProjectId         string            `mapstructure:"project_id"`
+	AccountFile       string            `mapstructure:"account_file"`
+	KeepOriginalImage bool              `mapstructure:"keep_input_artifact"`
```
`keep_input_artifact` is missing from the docs, except for that this looks good to go and that we have to sort out the tar & Go 1.10 issue.

rickard-von-essen commented:
This ready as soon as we sort out the GNU tar issue with Go 1.10.

seanmalloy commented:
Fixed issue where packer build was marked successful even when the GCE image creation operation failed.

Still facing issue with tar file format on Go 1.10. Will investigate further.

adarobin commented:
@seanmalloy If you can give me a macOS build I'd be happy to test/troubleshoot.  I'd love to have this post-processor for a project I'm working on.

seanmalloy commented:
@adarobin you can download the binaries from the below links. The packer binary compiled with Go 1.9 is working.

It is still not working on Go 1.10. The tar file created using Go 1.10 produces a GNU tar file, but Google cloud still does not accept it as a valid file. I'm still in the process of sorting out the problem with the tar file format generated by Go 1.10.

https://storage.googleapis.com/smalloy-sbx1-public/packer_go110_darwin_amd64.xz
https://storage.googleapis.com/smalloy-sbx1-public/packer_go19_darwin_amd64.xz

adarobin commented:
@seanmalloy I reproduced the issue with Go 1.10 I'll see if I can find what any differences in the file format.

Another issue I noticed is that `{{.BuildName}}` does not seem to work in the fields.

For example, I had:
```
{
  "type": "googlecompute-import",
  "project_id": "{{user `gcp_project_id`}}",
  "account_file": "{{user `gcp_account_file`}}",
  "bucket": "{{user `gcp_storage_bucket`}}",
  "image_name": "{{.BuildName}}-{{timestamp}}",
  "image_description": "{{.BuildName}}",
  "image_family": "rhel"
}
```
This failed, with the following:
```
==> GCP-RHEL6 (googlecompute-import): Uploading file GCP-RHEL6-compressed-image.tar.gz to GCS bucket redacted-bucket-name/packer-import-1529677804.tar.gz...
==> GCP-RHEL6 (googlecompute-import): Creating GCE image <no value>-1529677804...
==> GCP-RHEL6 (googlecompute-import): Error creating GCE image
Build 'GCP-RHEL6' errored: 1 error(s) occurred:

* Post-processor failed: googleapi: Error 400: Invalid value for field 'resource.name': '<no value>-1529677804'. Must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)', invalid

==> Some builds didn't complete successfully and had errors:
--> GCP-RHEL6: 1 error(s) occurred:

* Post-processor failed: googleapi: Error 400: Invalid value for field 'resource.name': '<no value>-1529677804'. Must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)', invalid

==> Builds finished but no artifacts were created.
```

I changed it to:
```
{
  "type": "googlecompute-import",
  "project_id": "{{user `gcp_project_id`}}",
  "account_file": "{{user `gcp_account_file`}}",
  "bucket": "{{user `gcp_storage_bucket`}}",
  "image_name": "test-{{timestamp}}",
  "image_description": "test",
  "image_family": "rhel"
}
```
And this works fine, other than the Go 1.10 tar issue.

rickard-von-essen commented:
@adarobin there is no variable `BuildName` here. You should use the function `{{ builder_name }}` instead, see https://www.packer.io/docs/templates/engine.html#functions

seanmalloy commented:
@adarobin yes I saw the same behaviour while trying to get the code to work. GCP just does not like the tar files generated by Go 1.10. Even when I added the code to tell Go 1.10 to explicitly create GNU tar files.

Accoring to the [GCP docs for image import](https://cloud.google.com/compute/docs/images/import-existing-image) the tar file must be in GNU tar format.

> The compressed file must be a .tar.gz file that uses gzip compression and the GNU tar format



Here is my update on this very exciting tar file format adventure ...

The `file` command is showing that the tar files created from Go 1.9.6 and 1.10.3 are both in GNU tar format. I `gunzip` both tar files and run the `file` command, and both show as GNU tar. This is good.

```
$ gunzip disk.raw.tar.gz
$ file disk.raw.tar
disk.raw.tar: POSIX tar archive (GNU)
```
When attempting to import the tar archive created by Go 1.10.3 into GCP the console shows the below error. This occurs when doing the import via the Go code in the googlecompute-import post-processor or when running a gcloud command to do the import. The tar files created by Go 1.9.6 work just fine without error. This is not cool.

![image](https://user-images.githubusercontent.com/685221/41805772-a72d1b1c-7675-11e8-84b8-ab31aded2fea.png)

Comparing the two different tar files using the `xxd` command I see some differences in the GNU tar file header generated by Go 1.9.6 and Go 1.10.3. Here is the best reference for the[ GNU tar header format](http://www.gnu.org/software/tar/manual/html_node/Standard.html) that I have been able to find.

In Go 1.10.3 it looks like lines 17 and 19 have the user and group in the header, and on lines 22 and 23 there are two additional timestamps(?) added in. These fields are not populated in the GNU tar file create by Go 1.9.6. Not sure if this is the problem, but these are the most obvious differences that I am noticing.

The `xxd` output for Go Lang 1.10.3 ...
```
$ xxd disk.raw.tar | cat -n
     1  0000000: 6469 736b 2e72 6177 0000 0000 0000 0000  disk.raw........
     2  0000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     3  0000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     4  0000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     5  0000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     6  0000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     7  0000060: 0000 0000 3030 3030 3634 3400 3030 3030  ....0000644.0000
     8  0000070: 3030 3000 3030 3030 3030 3000 8000 0000  000.0000000.....
     9  0000080: 0000 0003 a980 0000 3133 3331 3230 3736  ........13312076
    10  0000090: 3437 3300 3031 3434 3636 0020 3000 0000  473.014466. 0...
    11  00000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    12  00000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    13  00000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    14  00000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    15  00000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    16  00000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    17  0000100: 0075 7374 6172 2020 0072 6f6f 7400 0000  .ustar  .root...
    18  0000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    19  0000120: 0000 0000 0000 0000 0072 6f6f 7400 0000  .........root...
    20  0000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    21  0000140: 0000 0000 0000 0000 0030 3030 3030 3030  .........0000000
    22  0000150: 0030 3030 3030 3030 0031 3333 3132 3037  .0000000.1331207
    23  0000160: 3634 3637 0031 3333 3132 3037 3634 3733  6467.13312076473
    24  0000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    25  0000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    26  0000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    27  00001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    28  00001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    29  00001c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    30  00001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    31  00001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    32  00001f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    33  0000200: eb63 9010 8ed0 bc00 b0b8 0000 8ed8 8ec0  .c..............
    34  0000210: fbbe 007c bf00 06b9 0002 f3a4 ea21 0600  ...|.........!..
    35  0000220: 00be be07 3804 750b 83c6 1081 fefe 0775  ....8.u........u
    36  0000230: f3eb 16b4 02b0 01bb 007c b280 8a74 018b  .........|...t..
    37  0000240: 4c02 cd13 ea00 7c00 00eb fe00 0000 0000  L.....|.........
    38  0000250: 0000 0000 0000 0000 0000 0080 0100 0000  ................
    39  0000260: 0000 0000 fffa 9090 f6c2 8074 05f6 c270  ...........t...p
    40  0000270: 7402 b280 ea79 7c00 0031 c08e d88e d0bc  t....y|..1......
    41  0000280: 0020 fba0 647c 3cff 7402 88c2 52be 057c  . ..d|<.t...R..|
    42  0000290: 31c0 8944 0440 8844 ff89 4402 c704 1000  1..D.@.D..D.....
    43  00002a0: 668b 1e5c 7c66 895c 0866 8b1e 607c 6689  f..\|f.\.f..`|f.
    44  00002b0: 5c0c c744 0600 70b4 42cd 1372 05bb 0070  \..D..p.B..r...p
    45  00002c0: eb02 eb4b 601e b900 018e db31 f6bf 0080  ...K`......1....
    46  00002d0: 8ec6 fcf3 a51f 6160 b800 bbcd 1a66 85c0  ......a`.....f..
    47  00002e0: 751d b807 bbbf 0080 6631 f666 bb54 4350  u.......f1.f.TCP
    48  00002f0: 4166 b900 0200 0066 ba08 0000 00cd 1a61  Af.....f.......a
    49  0000300: ff26 5a7c be1f 7deb 03be 2e7d e834 00be  .&Z|..}....}.4..
    50  0000310: 337d e82e 00cd 18eb fe47 5255 4220 0047  3}.......GRUB .G
    51  0000320: 656f 6d00 4861 7264 2044 6973 6b00 5265  eom.Hard Disk.Re
    52  0000330: 6164 0020 4572 726f 720d 0a00 bb01 00b4  ad. Error.......
```

The `xxd` output for Go Lang 1.9.6 ...
```
     1  0000000: 6469 736b 2e72 6177 0000 0000 0000 0000  disk.raw........
     2  0000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     3  0000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     4  0000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     5  0000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     6  0000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     7  0000060: 0000 0000 3030 3030 3634 3400 3030 3030  ....0000644.0000
     8  0000070: 3030 3000 3030 3030 3030 3000 8000 0000  000.0000000.....
     9  0000080: 0000 0003 a980 0000 3133 3331 3233 3435  ........13312345
    10  0000090: 3135 3500 3031 3034 3735 0020 3000 0000  155.010475. 0...
    11  00000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    12  00000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    13  00000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    14  00000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    15  00000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    16  00000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    17  0000100: 0075 7374 6172 2020 0000 0000 0000 0000  .ustar  ........
    18  0000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    19  0000120: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    20  0000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    21  0000140: 0000 0000 0000 0000 0030 3030 3030 3030  .........0000000
    22  0000150: 0030 3030 3030 3030 0000 0000 0000 0000  .0000000........
    23  0000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    24  0000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    25  0000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    26  0000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    27  00001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    28  00001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    29  00001c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    30  00001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    31  00001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    32  00001f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    33  0000200: eb63 9010 8ed0 bc00 b0b8 0000 8ed8 8ec0  .c..............
    34  0000210: fbbe 007c bf00 06b9 0002 f3a4 ea21 0600  ...|.........!..
    35  0000220: 00be be07 3804 750b 83c6 1081 fefe 0775  ....8.u........u
    36  0000230: f3eb 16b4 02b0 01bb 007c b280 8a74 018b  .........|...t..
    37  0000240: 4c02 cd13 ea00 7c00 00eb fe00 0000 0000  L.....|.........
    38  0000250: 0000 0000 0000 0000 0000 0080 0100 0000  ................
    39  0000260: 0000 0000 fffa 9090 f6c2 8074 05f6 c270  ...........t...p
    40  0000270: 7402 b280 ea79 7c00 0031 c08e d88e d0bc  t....y|..1......
    41  0000280: 0020 fba0 647c 3cff 7402 88c2 52be 057c  . ..d|<.t...R..|
    42  0000290: 31c0 8944 0440 8844 ff89 4402 c704 1000  1..D.@.D..D.....
    43  00002a0: 668b 1e5c 7c66 895c 0866 8b1e 607c 6689  f..\|f.\.f..`|f.
    44  00002b0: 5c0c c744 0600 70b4 42cd 1372 05bb 0070  \..D..p.B..r...p
    45  00002c0: eb02 eb4b 601e b900 018e db31 f6bf 0080  ...K`......1....
    46  00002d0: 8ec6 fcf3 a51f 6160 b800 bbcd 1a66 85c0  ......a`.....f..
    47  00002e0: 751d b807 bbbf 0080 6631 f666 bb54 4350  u.......f1.f.TCP
    48  00002f0: 4166 b900 0200 0066 ba08 0000 00cd 1a61  Af.....f.......a
    49  0000300: ff26 5a7c be1f 7deb 03be 2e7d e834 00be  .&Z|..}....}.4..
    50  0000310: 337d e82e 00cd 18eb fe47 5255 4220 0047  3}.......GRUB .G
    51  0000320: 656f 6d00 4861 7264 2044 6973 6b00 5265  eom.Hard Disk.Re
    52  0000330: 6164 0020 4572 726f 720d 0a00 bb01 00b4  ad. Error.......
```

I think my next steps are to open a GitHub issue for Go Lang, and also open up a GCP support ticket to get additional information. Please stay tuned for updates in the near future. It might take me some time to get this sorted out.

adarobin commented:
@seanmalloy I definitely think this is a Go Lang issue, but I was able to get this working in my environment with this https://github.com/adarobin/packer/commit/f7f2f354e299f9767d90e5ad4132a329c4af8b3d

seanmalloy commented:
@adarobin I cherry-picked your commit into my branch and confirmed it is working.

Here is what the GNU tar file created by Go 1.10 now looks like ...

```
$ xxd disk.raw.tar | cat -n
     1  0000000: 6469 736b 2e72 6177 0000 0000 0000 0000  disk.raw........
     2  0000010: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     3  0000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     4  0000030: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     5  0000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     6  0000050: 0000 0000 0000 0000 0000 0000 0000 0000  ................
     7  0000060: 0000 0000 3030 3030 3634 3400 3030 3030  ....0000644.0000
     8  0000070: 3030 3000 3030 3030 3030 3000 8000 0000  000.0000000.....
     9  0000080: 0000 0003 a980 0000 3030 3030 3030 3030  ........00000000
    10  0000090: 3030 3000 3031 3232 3434 0020 3000 0000  000.012244. 0...
    11  00000a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    12  00000b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    13  00000c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    14  00000d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    15  00000e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    16  00000f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    17  0000100: 0075 7374 6172 2020 0072 6f6f 7400 0000  .ustar  .root...
    18  0000110: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    19  0000120: 0000 0000 0000 0000 0072 6f6f 7400 0000  .........root...
    20  0000130: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    21  0000140: 0000 0000 0000 0000 0030 3030 3030 3030  .........0000000
    22  0000150: 0030 3030 3030 3030 0000 0000 0000 0000  .0000000........
    23  0000160: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    24  0000170: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    25  0000180: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    26  0000190: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    27  00001a0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    28  00001b0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    29  00001c0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    30  00001d0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    31  00001e0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    32  00001f0: 0000 0000 0000 0000 0000 0000 0000 0000  ................
    33  0000200: eb63 9010 8ed0 bc00 b0b8 0000 8ed8 8ec0  .c..............
    34  0000210: fbbe 007c bf00 06b9 0002 f3a4 ea21 0600  ...|.........!..
    35  0000220: 00be be07 3804 750b 83c6 1081 fefe 0775  ....8.u........u
    36  0000230: f3eb 16b4 02b0 01bb 007c b280 8a74 018b  .........|...t..
    37  0000240: 4c02 cd13 ea00 7c00 00eb fe00 0000 0000  L.....|.........
    38  0000250: 0000 0000 0000 0000 0000 0080 0100 0000  ................
    39  0000260: 0000 0000 fffa 9090 f6c2 8074 05f6 c270  ...........t...p
    40  0000270: 7402 b280 ea79 7c00 0031 c08e d88e d0bc  t....y|..1......
    41  0000280: 0020 fba0 647c 3cff 7402 88c2 52be 057c  . ..d|<.t...R..|
    42  0000290: 31c0 8944 0440 8844 ff89 4402 c704 1000  1..D.@.D..D.....
    43  00002a0: 668b 1e5c 7c66 895c 0866 8b1e 607c 6689  f..\|f.\.f..`|f.
    44  00002b0: 5c0c c744 0600 70b4 42cd 1372 05bb 0070  \..D..p.B..r...p
    45  00002c0: eb02 eb4b 601e b900 018e db31 f6bf 0080  ...K`......1....
    46  00002d0: 8ec6 fcf3 a51f 6160 b800 bbcd 1a66 85c0  ......a`.....f..
    47  00002e0: 751d b807 bbbf 0080 6631 f666 bb54 4350  u.......f1.f.TCP
    48  00002f0: 4166 b900 0200 0066 ba08 0000 00cd 1a61  Af.....f.......a
    49  0000300: ff26 5a7c be1f 7deb 03be 2e7d e834 00be  .&Z|..}....}.4..
    50  0000310: 337d e82e 00cd 18eb fe47 5255 4220 0047  3}.......GRUB .G
    51  0000320: 656f 6d00 4861 7264 2044 6973 6b00 5265  eom.Hard Disk.Re
```

seanmalloy commented:
@mwhooker @SwampDragons and @rickard-von-essen thanks to @adarobin this code now works correctly on Go 1.10. I've removed WIP from the pull request title. I believe this code is ready for final review now. Let me know if any additional changes are needed.

SwampDragons commented:
Awesome sleuthing! Thanks, all!
